### PR TITLE
add fine alignment step to stabilize output pick point

### PR DESCRIPTION
**Pull request by lianjunli**
1. In recognition node, add fine alignment step to stabilize output pick point;
2. Lower z position of piston_rod to avoid the oscillating issue;
3. Change break beam location for better recognition result;

### CODE REVIEW

**Comment by jrgnicho** [comment]
@lianjunli thank you, I'll take a look and get back with you.

File: gilbreth_perception/config/model_list.yaml [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
-   pick_pose: [-0.025238, 0.080200, 0.668113, 0.0, 3.14159, 0.0] 
+   pick_pose: [-0.025238, 0.080200, 0.661113, 0.0, 3.14159, 0.0] 
  - name: gasket_part
    path: /model/gasket_part.pcd
-   pick_pose: [-0.116708, 0.093935, 0.674089, 0.0, 3.14159, 0.0] 
+   pick_pose: [-0.116708, 0.093935, 0.671089, 0.0, 3.14159, 0.0] 
  - name: disk_part 
    path: /model/disk_part.pcd
-   pick_pose: [-0.001646, 0.121226, 0.643602, 0.0, 3.14159, 0.0] 
+   pick_pose: [-0.001646, 0.121226, 0.646602, 0.0, 3.14159, 0.0]
```
Comment (jrgnicho) [inline gilbreth_perception/config/model_list.yaml:16]: How did you arrive to these numbers?

File: gilbreth_perception/config/parameters.yaml [hunk 2]
```diff
@@ -11,6 +11,8 @@ parameters:
   cg_size: 0.03
   cg_thresh: 8.0
   descr_dis_thrd: 0.1
+  #fine_align
+  iteration: 10
```
Comment (jrgnicho) [inline gilbreth_perception/config/parameters.yaml:15]: Does the alignment refinement iteration add significant time to the computation?

File: gilbreth_perception/src/recognition_node.cpp [hunk 3]
```diff
@@ -384,6 +387,22 @@ class RecognitionClass {
       pick_point.z = pick_pose[result.item_id][2];
       pick_point_cloud->push_back(pick_point);
       pcl::transformPointCloud(*pick_point_cloud, *rotated_pick_point_cloud, result.final_transformation);
+      // Use ICP to fine align model to scene
+      t_start = std::clock();
```
Comment (jrgnicho) [inline gilbreth_perception/src/recognition_node.cpp:391]: Why would you need to use ICP if you had already ran Correspondence Grouping?  Shouldn't you be able to use the transform from correspondence grouping to align the point cloud?

File: gilbreth_perception/src/recognition_node.cpp [hunk 4]
```diff
@@ -384,6 +387,22 @@ class RecognitionClass {
...
+      pcl::IterativeClosestPoint<pcl::PointXYZ, pcl::PointXYZ> icp;
+      icp.setMaximumIterations (iterations);
+      icp.setInputSource(rotated_model);
+      icp.setInputTarget(scene);
+      pcl::PointCloud<pcl::PointXYZ> Final;
```
Comment (jrgnicho) [inline gilbreth_perception/src/recognition_node.cpp]: Please follow variable naming conventions, use camelcase as described in the ros style guide.

File: gilbreth_perception/src/recognition_node.cpp [hunk 5]
```diff
@@ -384,6 +387,22 @@ class RecognitionClass {
...
+      icp.align(Final);
+      if (print_detailed_info){
+    	duration = (std::clock() - t_start) / (double)CLOCKS_PER_SEC;
+    	std::cerr << "align time is " << duration << " seconds." << std::endl;
```
Comment (jrgnicho) [inline gilbreth_perception/src/recognition_node.cpp]: Try using the ROS_INFO, ROS_ERROR, etc.  macros for printing to the console. You can use the ROS_INFO_STREAM and ROS_ERROR_STREAM if you'd like to use the **<<** stream insertion operators.

**Review by jrgnicho — COMMENTED** [review]
I looked at your changes and given the corresponding feedback, 
How did you test these changes, I'd like to use your testing setup in order to make sure the bug was fixed.  Thank you.

File: gilbreth_perception/config/model_list.yaml (hunk 1 above)
Comment (lianjunli) [reply gilbreth_perception/config/model_list.yaml:16]: I open the model PCD file with pcl_viewer, with the point picking option, use mouse to click on the point I'd like to be the pickup point, and the pcl_viewer return the coordination to me.

File: gilbreth_perception/config/parameters.yaml (hunk 2 above)
Comment (lianjunli) [reply gilbreth_perception/config/parameters.yaml:15]: On the server, it adds around 500ms on average.

File: gilbreth_perception/src/recognition_node.cpp (hunk 3 above)
Comment (lianjunli) [reply gilbreth_perception/src/recognition_node.cpp:391]: Correspondence grouping is kind of initial alignment, sometimes it is not very accurate, so the fine alignment step is needed.

File: gilbreth_perception/src/recognition_node.cpp (hunk 4 above)
Comment (lianjunli) [reply gilbreth_perception/src/recognition_node.cpp]: Will do that.

File: gilbreth_perception/src/recognition_node.cpp (hunk 5 above)
Comment (lianjunli) [reply gilbreth_perception/src/recognition_node.cpp]: Will do that.

File: gilbreth_perception/src/recognition_node.cpp (hunk 3 above)
Comment (jrgnicho) [reply gilbreth_perception/src/recognition_node.cpp:391]: Would it make more sense then to skip correspondence grouping and run ICP only?

File: gilbreth_perception/config/parameters.yaml (hunk 2 above)
Comment (jrgnicho) [reply gilbreth_perception/config/parameters.yaml:15]: That's not too bad

File: gilbreth_perception/src/recognition_node.cpp (hunk 3 above)
Comment (lianjunli) [reply gilbreth_perception/src/recognition_node.cpp:391]: I think I tried only ICP at the very beginning of this project, but the recognition result is not good, the fitness score was not a good metric for recognition, for example, there is a gear in the scene, the fitness score for matching model gear to scene gear is in the order of 10^-7,  the fitness score for matching model disk to scene gear is also in the order of 10^-7, there is a high chance that the system think the gear is a disk.
Also, there might be something wrong with my previous code, I can double check when I got time.

Comment (jrgnicho) [reply gilbreth_perception/src/recognition_node.cpp:391]: I see, lets use what you have right now and worry about those details later once we have running demo.

**Comment by jrgnicho** [comment]
@lianjunli this looks a lot better, I didn't run an exhaustive test but I found that the largest fluctuation was 0.008129 meters for the disk part. See the attached file for more results.  
[object_recognition_results2.txt](https://github.com/swri-robotics/gilbreth/files/1768854/object_recognition_results2.txt)

I think we need to improve the computational speed issue, perhaps we can cache some results (normal calculations, down-sampling, etc) and reuse them on subsequent operations.  I'll merge this PR and then we can brainstorm some more on further improvements.  Thanks for your contribution.

**Pull request merged**